File: web/src/lib/export/har.ts
/**
 * Build an HTTP Archive (HAR) 1.2 JSON object from flow data.
 *
 * Specification reference: http://www.softwareishard.com/blog/har-12-spec/
 *
 * The HAR is constructed entirely on the frontend from the data already
 * available via the MCP query tool — no backend changes are required.
 */

import type { FlowDetailResult, FlowEntry } from "../mcp/types.js";

// ---------------------------------------------------------------------------
// HAR type subset (only what we emit)
// ---------------------------------------------------------------------------

interface HarLog {
  log: {
    version: string;
    creator: { name: string; version: string };
    entries: HarEntry[];
  };
}

interface HarEntry {
  startedDateTime: string;
  time: number;
  request: HarRequest;
  response: HarResponse;
  cache: Record<string, never>;
  timings: { send: number; wait: number; receive: number };
}

interface HarRequest {
  method: string;
  url: string;
  httpVersion: string;
  headers: HarHeader[];
  queryString: HarQueryParam[];
  bodySize: number;
  headersSize: number;
  postData?: HarPostData;
}

interface HarResponse {
  status: number;
  statusText: string;
  httpVersion: string;
  headers: HarHeader[];
  content: HarContent;
  redirectURL: string;
  bodySize: number;
  headersSize: number;
}

interface HarHeader {
  name: string;
  value: string;
}

interface HarQueryParam {
  name: string;
  value: string;
}

interface HarPostData {
  mimeType: string;
  text: string;
}

interface HarContent {
  size: number;
  mimeType: string;
  text?: string;
  encoding?: string;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

const CREATOR = { name: "yorishiro-proxy", version: "1.0" };

/**
 * Build a complete HAR object from a single flow detail.
 */
export function buildHar(flow: FlowDetailResult): HarLog {
  return {
    log: {
      version: "1.2",
      creator: CREATOR,
      entries: [buildEntry(flow)],
    },
  };
}

/**
 * Build a HAR object from multiple flow entries.
 *
 * Because the list view ({@link FlowEntry}) contains less data than the
 * detail view, the generated entries will have fewer fields populated
 * (e.g. no bodies or full headers). This is still spec-compliant.
 */
export function buildHarFromList(flows: FlowEntry[]): HarLog {
  return {
    log: {
      version: "1.2",
      creator: CREATOR,
      entries: flows.map(buildEntryFromListItem),
    },
  };
}

/**
 * Trigger a file download of the HAR JSON in the browser.
 */
export function downloadHar(har: HarLog, filename: string): void {
  const json = JSON.stringify(har, null, 2);
  const blob = new Blob([json], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  try {
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
  } finally {
    URL.revokeObjectURL(url);
  }
}

// ---------------------------------------------------------------------------
// Entry builders
// ---------------------------------------------------------------------------

function buildEntry(flow: FlowDetailResult): HarEntry {
  const httpVersion = mapHttpVersion(flow.protocol);

  return {
    startedDateTime: flow.timestamp,
    time: flow.duration_ms,
    request: buildRequest(flow, httpVersion),
    response: buildResponse(flow, httpVersion),
    cache: {},
    timings: { send: -1, wait: flow.duration_ms, receive: -1 },
  };
}

function buildEntryFromListItem(flow: FlowEntry): HarEntry {
  const httpVersion = mapHttpVersion(flow.protocol);

  return {
    startedDateTime: flow.timestamp,
    time: flow.duration_ms,
    request: {
      method: flow.method ?? "GET",
      url: flow.url ?? "",
      httpVersion,
      headers: [],
      queryString: parseQueryString(flow.url),
      headersSize: -1,
      bodySize: -1,
    },
    response: {
      status: flow.status_code ?? 0,
      statusText: "",
      httpVersion,
      headers: [],
      content: { size: -1, mimeType: "" },
      redirectURL: "",
      headersSize: -1,
      bodySize: -1,
    },
    cache: {},
    timings: { send: -1, wait: flow.duration_ms, receive: -1 },
  };
}

// ---------------------------------------------------------------------------
// Request / Response helpers
// ---------------------------------------------------------------------------

function buildRequest(
  flow: FlowDetailResult,
  httpVersion: string,
): HarRequest {
  const headers = flattenHeaders(flow.request_headers);
  const contentType = getContentType(flow.request_headers);

  const req: HarRequest = {
    method: flow.method ?? "GET",
    url: flow.url ?? "",
    httpVersion,
    headers,
    queryString: parseQueryString(flow.url),
    headersSize: -1,
    bodySize: flow.request_body ? flow.request_body.length : 0,
  };

  if (flow.request_body) {
    req.postData = {
      mimeType: contentType,
      text: flow.request_body_encoding === "base64"
        ? flow.request_body  // Keep the base64 as-is; HAR supports encoding field
        : flow.request_body,
    };
  }

  return req;
}

function buildResponse(
  flow: FlowDetailResult,
  httpVersion: string,
): HarResponse {
  const headers = flattenHeaders(flow.response_headers);
  const contentType = getContentType(flow.response_headers);

  const content: HarContent = {
    size: flow.response_body ? flow.response_body.length : 0,
    mimeType: contentType,
  };

  if (flow.response_body) {
    if (flow.response_body_encoding === "base64") {
      content.text = flow.response_body;
      content.encoding = "base64";
    } else {
      content.text = flow.response_body;
    }
  }

  return {
    status: flow.response_status_code ?? 0,
    statusText: "",
    httpVersion,
    headers,
    content,
    redirectURL: getRedirectURL(flow.response_headers),
    headersSize: -1,
    bodySize: flow.response_body ? flow.response_body.length : 0,
  };
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function mapHttpVersion(protocol: string): string {
  switch (protocol) {
    case "HTTP/1.x":
      return "HTTP/1.1";
    case "HTTPS":
      return "HTTP/1.1";
    case "HTTP/2":
      return "h2";
    default:
      return "HTTP/1.1";
  }
}

/**
 * Flatten the multi-value headers map into a flat array of {name, value} pairs.
 */
function flattenHeaders(
  headers?: Record<string, string[]>,
): HarHeader[] {
  if (!headers) return [];
  const result: HarHeader[] = [];
  for (const [name, values] of Object.entries(headers)) {
    for (const value of values) {
      result.push({ name, value });
    }
  }
  return result;
}

/** Extract query string parameters from a URL. */
function parseQueryString(url?: string): HarQueryParam[] {
  if (!url) return [];
  try {
    const parsed = new URL(url);
    const params: HarQueryParam[] = [];
    parsed.searchParams.forEach((value, name) => {
      params.push({ name, value });
    });
    return params;
  } catch {
    return [];
  }
}

/** Extract the Content-Type value from a headers map. */
function getContentType(headers?: Record<string, string[]>): string {
  if (!headers) return "";
  for (const [name, values] of Object.entries(headers)) {
    if (name.toLowerCase() === "content-type" && values.length > 0) {
      return values[0];
    }
  }
  return "";
}

/** Extract the Location header for redirect URL. */
function getRedirectURL(headers?: Record<string, string[]>): string {
  if (!headers) return "";
  for (const [name, values] of Object.entries(headers)) {
    if (name.toLowerCase() === "location" && values.length > 0) {
      return values[0];
    }
  }
  return "";
}

File: internal/flow/har.go
package flow

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"mime"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"
)

// HARVersion is the HAR specification version produced by this exporter.
const HARVersion = "1.2"

// HAR is the top-level HTTP Archive structure.
type HAR struct {
	Log *HARLog `json:"log"`
}

// HARLog represents the root of the HAR data.
type HARLog struct {
	Version string      `json:"version"`
	Creator *HARCreator `json:"creator"`
	Entries []*HAREntry `json:"entries"`
}

// HARCreator identifies the tool that generated the HAR.
type HARCreator struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// HAREntry represents a single HTTP transaction.
type HAREntry struct {
	StartedDateTime string       `json:"startedDateTime"`
	Time            float64      `json:"time"`
	Request         *HARRequest  `json:"request"`
	Response        *HARResponse `json:"response"`
	Timings         *HARTimings  `json:"timings"`
	ServerIPAddress string       `json:"serverIPAddress,omitempty"`
	Connection      string       `json:"connection,omitempty"`

	// WebSocket custom field (Chrome DevTools convention).
	WebSocketMessages []*HARWebSocketMessage `json:"_webSocketMessages,omitempty"`
}

// HARRequest represents an HTTP request in HAR format.
type HARRequest struct {
	Method      string          `json:"method"`
	URL         string          `json:"url"`
	HTTPVersion string          `json:"httpVersion"`
	Headers     []*HARNameValue `json:"headers"`
	QueryString []*HARNameValue `json:"queryString"`
	PostData    *HARPostData    `json:"postData,omitempty"`
	HeadersSize int64           `json:"headersSize"`
	BodySize    int64           `json:"bodySize"`
}

// HARResponse represents an HTTP response in HAR format.
type HARResponse struct {
	Status      int             `json:"status"`
	StatusText  string          `json:"statusText"`
	HTTPVersion string          `json:"httpVersion"`
	Headers     []*HARNameValue `json:"headers"`
	Content     *HARContent     `json:"content"`
	HeadersSize int64           `json:"headersSize"`
	BodySize    int64           `json:"bodySize"`
}

// HARContent describes the response body content.
type HARContent struct {
	Size     int64  `json:"size"`
	MimeType string `json:"mimeType"`
	Text     string `json:"text,omitempty"`
	Encoding string `json:"encoding,omitempty"`
}

// HARPostData describes the request body.
type HARPostData struct {
	MimeType string `json:"mimeType"`
	Text     string `json:"text"`
}

// HARTimings contains timing information for the request.
type HARTimings struct {
	Send    float64 `json:"send"`
	Wait    float64 `json:"wait"`
	Receive float64 `json:"receive"`
}

// HARNameValue is a generic name-value pair used for headers and query parameters.
type HARNameValue struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// HARWebSocketMessage represents a WebSocket message (Chrome DevTools convention).
type HARWebSocketMessage struct {
	Type   string  `json:"type"`
	Time   float64 `json:"time"`
	Opcode int     `json:"opcode"`
	Data   string  `json:"data"`
}

// ExportHAR exports streams matching the filter to a HAR JSON file.
// It uses streaming JSON encoding to handle large stream sets efficiently.
// Returns the number of entries exported.
func ExportHAR(ctx context.Context, store Store, w io.Writer, opts ExportOptions, creatorVersion string) (int, error) {
	listOpts := StreamListOptions{
		Protocol:   opts.Filter.Protocol,
		URLPattern: opts.Filter.URLPattern,
	}

	streams, err := store.ListStreams(ctx, listOpts)
	if err != nil {
		return 0, fmt.Errorf("list streams for HAR export: %w", err)
	}

	if creatorVersion == "" {
		creatorVersion = "dev"
	}

	entries, exported, err := buildHAREntries(ctx, store, streams, opts)
	if err != nil {
		return exported, err
	}

	har := &HAR{
		Log: &HARLog{
			Version: HARVersion,
			Creator: &HARCreator{
				Name:    "yorishiro-proxy",
				Version: creatorVersion,
			},
			Entries: entries,
		},
	}

	enc := json.NewEncoder(w)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	if err := enc.Encode(har); err != nil {
		return exported, fmt.Errorf("encode HAR: %w", err)
	}

	return exported, nil
}

// buildHAREntries iterates over streams, applies filters, and converts each
// eligible stream to a HAR entry. Returns the entries, the count of exported
// streams, and any error encountered.
func buildHAREntries(ctx context.Context, store Store, streams []*Stream, opts ExportOptions) ([]*HAREntry, int, error) {
	// Initialize as empty slice so JSON serialization produces [] not null.
	entries := make([]*HAREntry, 0)
	exported := 0

	for _, st := range streams {
		if err := ctx.Err(); err != nil {
			return entries, exported, err
		}

		if opts.MaxFlows > 0 && exported >= opts.MaxFlows {
			break
		}

		if !harStreamIncluded(st, opts.Filter) {
			continue
		}

		flows, err := store.GetFlows(ctx, st.ID, FlowListOptions{})
		if err != nil {
			return entries, exported, fmt.Errorf("get flows for stream %s: %w", st.ID, err)
		}

		entry := convertStreamToHAREntry(st, flows, opts.IncludeBodies)
		if entry != nil {
			entries = append(entries, entry)
			exported++
		}
	}

	return entries, exported, nil
}

// harStreamIncluded returns true if the stream passes HAR-specific filters.
func harStreamIncluded(st *Stream, filter ExportFilter) bool {
	// Skip Raw TCP and gRPC binary frames per spec.
	if st.Protocol == "TCP" || st.Protocol == "gRPC" {
		return false
	}
	if filter.TimeAfter != nil && st.Timestamp.Before(*filter.TimeAfter) {
		return false
	}
	if filter.TimeBefore != nil && st.Timestamp.After(*filter.TimeBefore) {
		return false
	}
	return true
}

// convertStreamToHAREntry dispatches stream-to-HAR conversion based on protocol.
func convertStreamToHAREntry(st *Stream, flows []*Flow, includeBodies bool) *HAREntry {
	if st.Protocol == "WebSocket" {
		return streamToHARWebSocket(st, flows)
	}
	return streamToHAREntry(st, flows, includeBodies)
}

// streamToHAREntry converts a Stream and its flows to a HAR entry.
// Returns nil if the stream cannot be converted (e.g., missing request flow).
func streamToHAREntry(st *Stream, flows []*Flow, includeBodies bool) *HAREntry {
	var reqFlow, respFlow *Flow
	for _, f := range flows {
		switch f.Direction {
		case "send":
			if reqFlow == nil {
				reqFlow = f
			}
		case "receive":
			if respFlow == nil {
				respFlow = f
			}
		}
	}

	if reqFlow == nil {
		return nil
	}

	entry := &HAREntry{
		StartedDateTime: st.Timestamp.UTC().Format("2006-01-02T15:04:05.000Z"),
		Time:            float64(st.Duration.Milliseconds()),
		Request:         buildHARRequest(reqFlow, st.Protocol, includeBodies),
		Response:        buildHARResponse(respFlow, st.Protocol, includeBodies),
		Timings:         buildHARTimings(st),
	}

	if st.ConnInfo != nil {
		entry.ServerIPAddress = extractIP(st.ConnInfo.ServerAddr)
		entry.Connection = st.ConnID
	}

	return entry
}

// streamToHARWebSocket converts a WebSocket stream to a HAR entry with
// the _webSocketMessages custom field.
func streamToHARWebSocket(st *Stream, flows []*Flow) *HAREntry {
	// Find the initial HTTP upgrade request/response.
	var reqFlow, respFlow *Flow
	var wsMessages []*HARWebSocketMessage

	for _, f := range flows {
		switch f.Direction {
		case "send":
			if reqFlow == nil && f.Method != "" {
				reqFlow = f
			} else {
				wsMessages = append(wsMessages, &HARWebSocketMessage{
					Type:   "send",
					Time:   float64(f.Timestamp.UnixMilli()) / 1000.0,
					Opcode: parseOpcode(f.Metadata),
					Data:   wsFlowData(f),
				})
			}
		case "receive":
			if respFlow == nil && f.StatusCode != 0 {
				respFlow = f
			} else {
				wsMessages = append(wsMessages, &HARWebSocketMessage{
					Type:   "receive",
					Time:   float64(f.Timestamp.UnixMilli()) / 1000.0,
					Opcode: parseOpcode(f.Metadata),
					Data:   wsFlowData(f),
				})
			}
		}
	}

	// Build the entry from the upgrade request if available.
	entry := &HAREntry{
		StartedDateTime:   st.Timestamp.UTC().Format("2006-01-02T15:04:05.000Z"),
		Time:              float64(st.Duration.Milliseconds()),
		Request:           buildHARRequest(reqFlow, st.Protocol, false),
		Response:          buildHARResponse(respFlow, st.Protocol, false),
		Timings:           buildHARTimings(st),
		WebSocketMessages: wsMessages,
	}

	if st.ConnInfo != nil {
		entry.ServerIPAddress = extractIP(st.ConnInfo.ServerAddr)
		entry.Connection = st.ConnID
	}

	return entry
}

// buildHARRequest constructs a HARRequest from a send flow.
func buildHARRequest(f *Flow, protocol string, includeBodies bool) *HARRequest {
	if f == nil {
		return &HARRequest{
			Method:      "GET",
			URL:         "",
			HTTPVersion: protocolToHTTPVersion(protocol),
			Headers:     []*HARNameValue{},
			QueryString: []*HARNameValue{},
			HeadersSize: -1,
			BodySize:    -1,
		}
	}

	reqURL := ""
	var queryParams []*HARNameValue
	if f.URL != nil {
		reqURL = f.URL.String()
		queryParams = urlQueryToHAR(f.URL.Query())
	}
	if queryParams == nil {
		queryParams = []*HARNameValue{}
	}

	method := f.Method
	if method == "" {
		method = "GET"
	}

	req := &HARRequest{
		Method:      method,
		URL:         reqURL,
		HTTPVersion: protocolToHTTPVersion(protocol),
		Headers:     headersToHAR(f.Headers),
		QueryString: queryParams,
		HeadersSize: -1,
		BodySize:    int64(len(f.Body)),
	}

	if includeBodies && len(f.Body) > 0 {
		mimeType := "application/octet-stream"
		if ct := headerValue(f.Headers, "Content-Type"); ct != "" {
			mimeType = ct
		}
		req.PostData = &HARPostData{
			MimeType: mimeType,
			Text:     bodyToText(f.Body),
		}
	}

	return req
}

// buildHARResponse constructs a HARResponse from a receive flow.
func buildHARResponse(f *Flow, protocol string, includeBodies bool) *HARResponse {
	if f == nil {
		return &HARResponse{
			Status:      0,
			StatusText:  "",
			HTTPVersion: protocolToHTTPVersion(protocol),
			Headers:     []*HARNameValue{},
			Content: &HARContent{
				Size:     0,
				MimeType: "",
			},
			HeadersSize: -1,
			BodySize:    -1,
		}
	}

	resp := &HARResponse{
		Status:      f.StatusCode,
		StatusText:  statusText(f.StatusCode),
		HTTPVersion: protocolToHTTPVersion(protocol),
		Headers:     headersToHAR(f.Headers),
		HeadersSize: -1,
		BodySize:    int64(len(f.Body)),
	}

	mimeType := "application/octet-stream"
	if ct := headerValue(f.Headers, "Content-Type"); ct != "" {
		mimeType = ct
	}

	content := &HARContent{
		Size:     int64(len(f.Body)),
		MimeType: mimeType,
	}

	if includeBodies && len(f.Body) > 0 {
		if isBinaryContent(mimeType, f.Body) {
			content.Text = base64.StdEncoding.EncodeToString(f.Body)
			content.Encoding = "base64"
		} else {
			content.Text = string(f.Body)
		}
	}

	resp.Content = content
	return resp
}

// buildHARTimings constructs HAR timings from stream timing data.
func buildHARTimings(st *Stream) *HARTimings {
	t := &HARTimings{
		Send:    -1,
		Wait:    -1,
		Receive: -1,
	}
	if st.SendMs != nil {
		t.Send = float64(*st.SendMs)
	}
	if st.WaitMs != nil {
		t.Wait = float64(*st.WaitMs)
	}
	if st.ReceiveMs != nil {
		t.Receive = float64(*st.ReceiveMs)
	}
	return t
}

// headersToHAR converts HTTP headers to HAR name-value pairs.
func headersToHAR(headers map[string][]string) []*HARNameValue {
	if len(headers) == 0 {
		return []*HARNameValue{}
	}

	// Sort header names for deterministic output.
	names := make([]string, 0, len(headers))
	for name := range headers {
		names = append(names, name)
	}
	sort.Strings(names)

	var result []*HARNameValue
	for _, name := range names {
		for _, value := range headers[name] {
			result = append(result, &HARNameValue{
				Name:  name,
				Value: value,
			})
		}
	}
	return result
}

// urlQueryToHAR converts URL query parameters to HAR name-value pairs.
func urlQueryToHAR(values url.Values) []*HARNameValue {
	if len(values) == 0 {
		return []*HARNameValue{}
	}

	// Sort keys for deterministic output.
	keys := make([]string, 0, len(values))
	for k := range values {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var result []*HARNameValue
	for _, k := range keys {
		for _, v := range values[k] {
			result = append(result, &HARNameValue{
				Name:  k,
				Value: v,
			})
		}
	}
	return result
}

// headerValue returns the first value for the given header name (case-insensitive).
func headerValue(headers map[string][]string, name string) string {
	if headers == nil {
		return ""
	}
	// Try exact match first.
	if vals, ok := headers[name]; ok && len(vals) > 0 {
		return vals[0]
	}
	// Case-insensitive fallback.
	lowerName := strings.ToLower(name)
	for k, vals := range headers {
		if strings.ToLower(k) == lowerName && len(vals) > 0 {
			return vals[0]
		}
	}
	return ""
}

// protocolToHTTPVersion maps proxy protocol names to HAR httpVersion strings.
func protocolToHTTPVersion(protocol string) string {
	switch protocol {
	case "HTTP/1.x":
		return "HTTP/1.1"
	case "HTTPS":
		return "HTTP/1.1"
	case "HTTP/2":
		return "h2"
	case "WebSocket":
		return "HTTP/1.1"
	default:
		if strings.HasPrefix(protocol, "SOCKS5+") {
			return protocolToHTTPVersion(strings.TrimPrefix(protocol, "SOCKS5+"))
		}
		return "HTTP/1.1"
	}
}

// isBinaryContent determines if content should be base64-encoded in HAR.
func isBinaryContent(mimeType string, body []byte) bool {
	// Parse the media type, ignoring parameters.
	mediaType, _, _ := mime.ParseMediaType(mimeType)
	if mediaType == "" {
		mediaType = mimeType
	}

	// Text-based MIME types.
	if strings.HasPrefix(mediaType, "text/") {
		return false
	}
	textTypes := []string{
		"application/json",
		"application/xml",
		"application/xhtml+xml",
		"application/javascript",
		"application/x-javascript",
		"application/ecmascript",
		"application/x-www-form-urlencoded",
		"application/soap+xml",
		"application/graphql",
		"application/ld+json",
		"application/manifest+json",
	}
	for _, t := range textTypes {
		if mediaType == t {
			return false
		}
	}
	// Suffix-based detection (e.g., application/vnd.api+json).
	if strings.HasSuffix(mediaType, "+json") || strings.HasSuffix(mediaType, "+xml") {
		return false
	}

	// Known binary MIME prefixes — always base64-encode regardless of body content.
	binaryPrefixes := []string{"image/", "audio/", "video/", "font/"}
	for _, prefix := range binaryPrefixes {
		if strings.HasPrefix(mediaType, prefix) {
			return true
		}
	}

	// Check if the body is valid UTF-8 without control characters.
	if utf8.Valid(body) {
		return false
	}

	return true
}

// extractIP extracts the IP address from a host:port string.
func extractIP(addr string) string {
	if addr == "" {
		return ""
	}
	// Try to split host:port.
	if idx := strings.LastIndex(addr, ":"); idx >= 0 {
		host := addr[:idx]
		// Handle IPv6 bracket notation.
		if strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]") {
			return host[1 : len(host)-1]
		}
		return host
	}
	return addr
}

// parseOpcode extracts the WebSocket opcode from flow metadata.
func parseOpcode(metadata map[string]string) int {
	if metadata == nil {
		return 0
	}
	if s, ok := metadata["opcode"]; ok {
		if v, err := strconv.Atoi(s); err == nil {
			return v
		}
	}
	return 0
}

// wsFlowData returns the WebSocket flow data as a string.
// Binary data is base64-encoded.
func wsFlowData(f *Flow) string {
	if len(f.Body) == 0 {
		return ""
	}
	if utf8.Valid(f.Body) {
		return string(f.Body)
	}
	return base64.StdEncoding.EncodeToString(f.Body)
}

// bodyToText converts a request body to text for HAR postData.
// Binary data is base64-encoded.
func bodyToText(body []byte) string {
	if utf8.Valid(body) {
		return string(body)
	}
	return base64.StdEncoding.EncodeToString(body)
}

// statusText returns the reason phrase for an HTTP status code.
// This is a local copy to avoid an import cycle with httputil (which imports flow).
// The canonical source is httputil.StatusText. Keep in sync when updating.
var httpStatusTexts = map[int]string{
	100: "Continue", 101: "Switching Protocols", 102: "Processing", 103: "Early Hints",

	200: "OK", 201: "Created", 202: "Accepted", 203: "Non-Authoritative Information",
	204: "No Content", 205: "Reset Content", 206: "Partial Content",
	207: "Multi-Status", 208: "Already Reported", 226: "IM Used",

	300: "Multiple Choices", 301: "Moved Permanently", 302: "Found", 303: "See Other",
	304: "Not Modified", 305: "Use Proxy", 307: "Temporary Redirect", 308: "Permanent Redirect",

	400: "Bad Request", 401: "Unauthorized", 402: "Payment Required", 403: "Forbidden",
	404: "Not Found", 405: "Method Not Allowed", 406: "Not Acceptable",
	407: "Proxy Authentication Required", 408: "Request Timeout", 409: "Conflict",
	410: "Gone", 411: "Length Required", 412: "Precondition Failed",
	413: "Request Entity Too Large", 414: "Request URI Too Long",
	415: "Unsupported Media Type", 416: "Requested Range Not Satisfiable",
	417: "Expectation Failed", 418: "I'm a teapot", 421: "Misdirected Request",
	422: "Unprocessable Entity", 423: "Locked", 424: "Failed Dependency",
	425: "Too Early", 426: "Upgrade Required", 428: "Precondition Required",
	429: "Too Many Requests", 431: "Request Header Fields Too Large",
	451: "Unavailable For Legal Reasons",

	500: "Internal Server Error", 501: "Not Implemented", 502: "Bad Gateway",
	503: "Service Unavailable", 504: "Gateway Timeout",
	505: "HTTP Version Not Supported", 506: "Variant Also Negotiates",
	507: "Insufficient Storage", 508: "Loop Detected", 510: "Not Extended",
	511: "Network Authentication Required",
}

func statusText(code int) string {
	return httpStatusTexts[code]
}
